Make creating directories lazy in RootWorkspace

mgrzaslewicz opened the pull request:
That stops from creating directories just by creating instance of RootWorkspace

dagguh commented on src/main/kotlin/com/atlassian/performance/tools/workspace/api/RootWorkspace.kt:
```diff
@@ -19,20 +19,24 @@ class RootWorkspace(
-        .toAbsolutePath()
-        .normalize()
-        .ensureDirectory()
+    val directory: Path by lazy {
```
Best to split the `lazy` commit from the `AssertJ` commit

dagguh commented on src/main/kotlin/com/atlassian/performance/tools/workspace/api/RootWorkspace.kt:
```diff
@@ -19,20 +19,24 @@ class RootWorkspace(
-            ?: LocalDateTime.now()
-                .format(ISO_LOCAL_DATE_TIME)
-    )
+    val currentTask: TaskWorkspace by lazy {
```
Once we have a lazy `Path`, here we have a lazy `TaskWorkspace`. Let's make it consistent, e.g. lazy `Path` fields.

mgrzaslewicz replied:
Which fields do you mean? The only one not lazy in this class is `private val parent` but creating it has no side effect so it can stay as it is

dagguh replied:
Ah, so the pattern is: `lazy` around `ensureDirectory`?

dagguh replied:
I mean, `RootWorkspace` represents its `directory` field. Same for `TaskWorkspace`.
But the root one ensures its existence, but the task one doesn't. `RootWorkspace` kinda "stole" the `ensureDirectory`, which should have been inside `TaskWorkspace`.
PS. same goes for `TestWorkspace`.

mgrzaslewicz replied:
I've unified the approach so every class takes care of its own directory creation lazily
- `TestWorkspace`
- `TaskWorkspace`
- `RootWorkspace`

dagguh replied:
That makes a lot of sense! Sadly, we're caught by surprise by [our tech debt here](https://github.com/atlassian/workspace/pull/11#discussion_r1045950747).

dagguh commented on CHANGELOG.md:
```diff
@@ -23,6 +23,9 @@ Dropping a requirement of a major version of a dependency is a new contract.
+### Fixed
+- Stop creating directories just by creating instance of `RootWorkspace`, make it lazy.
+
 ## [2.0.2] - 2019-11-04
```
It's been a looong time since last change :)
We can port this to GHA if TravisCI is not cooperating

dagguh commented on src/test/kotlin/com/atlassian/performance/tools/report/RootWorkspaceTest.kt:
```diff
@@ -0,0 +1,30 @@
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.Test
+
+class RootWorkspaceTest {
```
Tests, nice!

dagguh commented on src/test/kotlin/com/atlassian/performance/tools/report/RootWorkspaceTest.kt:
```diff
@@ -0,0 +1,30 @@
+    @Test
+    fun shouldCreatingNewInstanceHaveNoSideEffect() {
+        // given
+        val tempDir = Files.newTemporaryFolder()
```
It doesn't matter that it's temporary:
```suggestion
        val dir = Files.newTemporaryFolder()
```
Or:
```suggestion
        val parent = Files.newTemporaryFolder()
```

mgrzaslewicz-atlassian committed:
Replace hamcrest with assertj in unit tests

dagguh commented on src/test/kotlin/com/atlassian/performance/tools/report/RootWorkspaceTest.kt:
```diff
@@ -0,0 +1,30 @@
+        // given
+        val tempDir = Files.newTemporaryFolder()
+        // when
+        RootWorkspace(parent = tempDir.toPath()).currentTask
```
You can use `java.nio.file.Files.createTempDirectory` and have it as a `Path` from the get-go.

mgrzaslewicz replied:
I know and I picked assertj one as it's easier to use and read the code

mgrzaslewicz-atlassian committed:
Make creating directories lazy in RootWorkspace

That stops from creating directories just by creating instance of RootWorkspace

dagguh commented on src/test/kotlin/com/atlassian/performance/tools/report/RootWorkspaceTest.kt:
```diff
@@ -0,0 +1,30 @@
+        // when
+        RootWorkspace(parent = tempDir.toPath()).currentTask
+        // then
+        assertThat(tempDir).isNotEmptyDirectory()
```
What does it contain in this case? Just the `RootWorkspace.directory` or `RootWorkspace.currentTask.directory`?
If the problem is "sometimes we'll have empty directories", then this still happens here, right?

mgrzaslewicz replied:
`RootWorkspace.directory` until `TaskWorkspace.directory` is read

mgrzaslewicz replied:
I've changed the test to reflect better what's actually tested

dagguh commented on src/main/kotlin/com/atlassian/performance/tools/workspace/api/TestWorkspace.kt:
```diff
@@ -1,10 +1,28 @@
 /**
  * Contains files necessary to make the test work.
  */
-data class TestWorkspace(
```
Unfortunately Kotlin also generates:
* a `copy` method, to be called directly.
* `component1` method for [destructuring declarations](https://kotlinlang.org/docs/destructuring-declarations.html)

<details>
<summary>TestWorkspace bytecode</summary>

```java
// class version 50.0 (50)
// access flags 0x31
public final class com/atlassian/performance/tools/workspace/api/TestWorkspace {

  // compiled from: TestWorkspace.kt

  @Lkotlin/Metadata;(mv={1, 1, 13}, bv={1, 0, 3}, k=1, d1={"\u0000&\n\u0002\u0018\u0002\n\u0002\u0010\u0000\n\u0000\n\u0002\u0018\u0002\n\u0002\u0008\u0006\n\u0002\u0010\u000b\n\u0002\u0008\u0002\n\u0002\u0010\u0008\n\u0000\n\u0002\u0010\u000e\n\u0000\u0008\u0086\u0008\u0018\u00002\u00020\u0001B\r\u0012\u0006\u0010\u0002\u001a\u00020\u0003\u00a2\u0006\u0002\u0010\u0004J\u0009\u0010\u0007\u001a\u00020\u0003H\u00c6\u0003J\u0013\u0010\u0008\u001a\u00020\u00002\u0008\u0008\u0002\u0010\u0002\u001a\u00020\u0003H\u00c6\u0001J\u0013\u0010\u0009\u001a\u00020\n2\u0008\u0010\u000b\u001a\u0004\u0018\u00010\u0001H\u00d6\u0003J\u0009\u0010\u000c\u001a\u00020\rH\u00d6\u0001J\u0009\u0010\u000e\u001a\u00020\u000fH\u00d6\u0001R\u0011\u0010\u0002\u001a\u00020\u0003\u00a2\u0006\u0008\n\u0000\u001a\u0004\u0008\u0005\u0010\u0006\u00a8\u0006\u0010"}, d2={"Lcom/atlassian/performance/tools/workspace/api/TestWorkspace;", "", "directory", "Ljava/nio/file/Path;", "(Ljava/nio/file/Path;)V", "getDirectory", "()Ljava/nio/file/Path;", "component1", "copy", "equals", "", "other", "hashCode", "", "toString", "", "workspace"})

  // access flags 0x12
  private final Ljava/nio/file/Path; directory
  @Lorg/jetbrains/annotations/NotNull;() // invisible

  // access flags 0x11
  public final getDirectory()Ljava/nio/file/Path;
  @Lorg/jetbrains/annotations/NotNull;() // invisible
   L0
    LINENUMBER 9 L0
    ALOAD 0
    GETFIELD com/atlassian/performance/tools/workspace/api/TestWorkspace.directory : Ljava/nio/file/Path;
    ARETURN
   L1
    LOCALVARIABLE this Lcom/atlassian/performance/tools/workspace/api/TestWorkspace; L0 L1 0
    MAXSTACK = 1
    MAXLOCALS = 1

  // access flags 0x1
  public <init>(Ljava/nio/file/Path;)V
    // annotable parameter count: 1 (invisible)
    @Lorg/jetbrains/annotations/NotNull;() // invisible, parameter 0
   L0
    ALOAD 1
    LDC "directory"
    INVOKESTATIC kotlin/jvm/internal/Intrinsics.checkParameterIsNotNull (Ljava/lang/Object;Ljava/lang/String;)V
   L1
    LINENUMBER 8 L1
    ALOAD 0
    INVOKESPECIAL java/lang/Object.<init> ()V
    ALOAD 0
    ALOAD 1
    PUTFIELD com/atlassian/performance/tools/workspace/api/TestWorkspace.directory : Ljava/nio/file/Path;
    RETURN
   L2
    LOCALVARIABLE this Lcom/atlassian/performance/tools/workspace/api/TestWorkspace; L0 L2 0
    LOCALVARIABLE directory Ljava/nio/file/Path; L0 L2 1
    MAXSTACK = 2
    MAXLOCALS = 2

  // access flags 0x11
  public final component1()Ljava/nio/file/Path;
  @Lorg/jetbrains/annotations/NotNull;() // invisible
   L0
    ALOAD 0
    GETFIELD com/atlassian/performance/tools/workspace/api/TestWorkspace.directory : Ljava/nio/file/Path;
    ARETURN
   L1
    LOCALVARIABLE this Lcom/atlassian/performance/tools/workspace/api/TestWorkspace; L0 L1 0
    MAXSTACK = 1
    MAXLOCALS = 1

  // access flags 0x11
  public final copy(Ljava/nio/file/Path;)Lcom/atlassian/performance/tools/workspace/api/TestWorkspace;
  @Lorg/jetbrains/annotations/NotNull;() // invisible
    // annotable parameter count: 1 (invisible)
    @Lorg/jetbrains/annotations/NotNull;() // invisible, parameter 0
   L0
    ALOAD 1
    LDC "directory"
    INVOKESTATIC kotlin/jvm/internal/Intrinsics.checkParameterIsNotNull (Ljava/lang/Object;Ljava/lang/String;)V
    NEW com/atlassian/performance/tools/workspace/api/TestWorkspace
    DUP
    ALOAD 1
    INVOKESPECIAL com/atlassian/performance/tools/workspace/api/TestWorkspace.<init> (Ljava/nio/file/Path;)V
    ARETURN
   L1
    LOCALVARIABLE this Lcom/atlassian/performance/tools/workspace/api/TestWorkspace; L0 L1 0
    LOCALVARIABLE directory Ljava/nio/file/Path; L0 L1 1
    MAXSTACK = 3
    MAXLOCALS = 2

  // access flags 0x1009
  public static synthetic copy$default(Lcom/atlassian/performance/tools/workspace/api/TestWorkspace;Ljava/nio/file/Path;ILjava/lang/Object;)Lcom/atlassian/performance/tools/workspace/api/TestWorkspace;
  @Lorg/jetbrains/annotations/NotNull;() // invisible
    ILOAD 2
    ICONST_1
    IAND
    IFEQ L0
    ALOAD 0
    GETFIELD com/atlassian/performance/tools/workspace/api/TestWorkspace.directory : Ljava/nio/file/Path;
    ASTORE 1
   L0
   FRAME SAME
    ALOAD 0
    ALOAD 1
    INVOKEVIRTUAL com/atlassian/performance/tools/workspace/api/TestWorkspace.copy (Ljava/nio/file/Path;)Lcom/atlassian/performance/tools/workspace/api/TestWorkspace;
    ARETURN
    MAXSTACK = 2
    MAXLOCALS = 4

  // access flags 0x1
  public toString()Ljava/lang/String;
  @Lorg/jetbrains/annotations/NotNull;() // invisible
    NEW java/lang/StringBuilder
    DUP
    INVOKESPECIAL java/lang/StringBuilder.<init> ()V
    LDC "TestWorkspace(directory="
    INVOKEVIRTUAL java/lang/StringBuilder.append (Ljava/lang/String;)Ljava/lang/StringBuilder;
    ALOAD 0
    GETFIELD com/atlassian/performance/tools/workspace/api/TestWorkspace.directory : Ljava/nio/file/Path;
    INVOKEVIRTUAL java/lang/StringBuilder.append (Ljava/lang/Object;)Ljava/lang/StringBuilder;
    LDC ")"
    INVOKEVIRTUAL java/lang/StringBuilder.append (Ljava/lang/String;)Ljava/lang/StringBuilder;
    INVOKEVIRTUAL java/lang/StringBuilder.toString ()Ljava/lang/String;
    ARETURN
    MAXSTACK = 2
    MAXLOCALS = 1

  // access flags 0x1
  public hashCode()I
    ALOAD 0
    GETFIELD com/atlassian/performance/tools/workspace/api/TestWorkspace.directory : Ljava/nio/file/Path;
    DUP
    IFNULL L0
    INVOKEVIRTUAL java/lang/Object.hashCode ()I
    GOTO L1
   L0
   FRAME SAME1 java/nio/file/Path
    POP
    ICONST_0
   L1
   FRAME SAME1 I
    IRETURN
    MAXSTACK = 2
    MAXLOCALS = 1

  // access flags 0x1
  public equals(Ljava/lang/Object;)Z
    // annotable parameter count: 1 (invisible)
    @Lorg/jetbrains/annotations/Nullable;() // invisible, parameter 0
    ALOAD 0
    ALOAD 1
    IF_ACMPEQ L0
    ALOAD 1
    INSTANCEOF com/atlassian/performance/tools/workspace/api/TestWorkspace
    IFEQ L1
    ALOAD 1
    CHECKCAST com/atlassian/performance/tools/workspace/api/TestWorkspace
    ASTORE 2
    ALOAD 0
    GETFIELD com/atlassian/performance/tools/workspace/api/TestWorkspace.directory : Ljava/nio/file/Path;
    ALOAD 2
    GETFIELD com/atlassian/performance/tools/workspace/api/TestWorkspace.directory : Ljava/nio/file/Path;
    INVOKESTATIC kotlin/jvm/internal/Intrinsics.areEqual (Ljava/lang/Object;Ljava/lang/Object;)Z
    IFEQ L1
   L0
   FRAME SAME
    ICONST_1
    IRETURN
   L1
   FRAME SAME
    ICONST_0
    IRETURN
    MAXSTACK = 2
    MAXLOCALS = 3
}
```
</details>

These are the compatibility pains that we learned too late. It's hard to know whether they're used somewhere or not, so we're pessimistic by default and treat this as API.
PS. `copy` is actually called in our internal repos, look for `val firstPhaseWorkspace = workspace.copy(`.

Options:
* try to add `lazy` without dropping `data class`
* try to reimplement those generated methods manually
* pay off the tech debt, break compatibility, remove the `data class`, release a major version
    * there's gonna be some pain with that, but I can help

mgrzaslewicz replied:
> try to add lazy without dropping data class

Unfortunately data class must contain val/var in the constructor and that's the reason I've dropped `data`

> pay off the tech debt, break compatibility, remove the data class, release a major version 

I think it's the way to go

pczuj replied:
2nd option will be good for visibility of compatibility maintenance in the future changes, i.e. the methods existance would be explicit. Would be cool to have that.

pczuj replied:
Alternatively we could keep the data class as internal component of the new `TestWorkspace` implementation and we could delegate the lost implicit methods to that internal data class, while keeping the side effect as result of new getter (post delegation to data class).

Not sure if the complexity is worth though. Possibly it's better to just reimplement all of the methods from scratch.
